### PR TITLE
pybind/rados: add application_metadata_get

### PR DESCRIPTION
**Pull request by Songweibin**
Signed-off-by: songweibin <song.weibin@zte.com.cn>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`

</details>

### CODE REVIEW

File: src/pybind/rados/rados.pyx [hunk 1]
```diff
@@ -3937,6 +3937,44 @@ returned %d, but should return zero on success." % (self.name, ret))
...
+
+        :raises: :class:`Error`
+        """
+
+        app_name =  cstr(app_name, 'app_name')
+        key = cstr(key, 'key')
+        cdef:
+            char *_app_name = app_name
+            char *_key = key
+            size_t size = 4096
```
Comment (xiexingguo) [inline src/pybind/rados/rados.pyx]: the max_application_data_length is hard-coded at 128, so 128 + 1 would suffice, see https://github.com/ceph/ceph/blob/master/src/mon/OSDMonitor.cc#L231

Comment (Songweibin) [reply src/pybind/rados/rados.pyx]: done, thx.

File: src/pybind/rados/rados.pyx [hunk 2]
```diff
@@ -3937,6 +3937,44 @@ returned %d, but should return zero on success." % (self.name, ret))
...
+            while True:
+                value = <char *>realloc_chk(value, size)
+                with nogil:
+                    ret = rados_application_metadata_get(self.io, _app_name,
+                                                         _key, value, &size)
+                if ret != -errno.ERANGE:
+                    break
+            if ret == -errno.ENOENT:
+                raise KeyError('no metadata %s for application %s' % (key, _app_name))
+            if ret != 0:
```
Comment (xiexingguo) [inline src/pybind/rados/rados.pyx]: elif ?

Comment (Songweibin) [reply src/pybind/rados/rados.pyx]: updated, thx.

**Comment by xiexingguo** [comment]
retest this please

File: src/pybind/rados/rados.pyx [hunk 3]
```diff
@@ -3937,6 +3937,44 @@ returned %d, but should return zero on success." % (self.name, ret))
...
+                with nogil:
+                    ret = rados_application_metadata_get(self.io, _app_name,
+                                                         _key, value, &size)
+                if ret != -errno.ERANGE:
+                    break
+            if ret == -errno.ENOENT:
+                raise KeyError('no metadata %s for application %s' % (key, _app_name))
+            elif ret != 0:
+                raise make_ex(ret, 'error getting metadata %s for application %s' %
+                              (key, _app_name,))
```
Comment (xiexingguo) [inline src/pybind/rados/rados.pyx]: could you drop the extra trailing `,`?

Comment (Songweibin) [reply src/pybind/rados/rados.pyx]: sorry ...... updated.

**Comment by xiexingguo** [comment]
retest this please

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2019-09-25_01:32:27-rados-wip-kefu-testing-2019-09-24-1556-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2019-09-27_03:53:11-rados-wip-kefu-testing-2019-09-24-1556-distro-basic-smithi/

**Pull request merged**